### DISCUSSION TITLE
Adding Custom Functions to Liquidsoap Server

### DISCUSSION THREAD

**Discussion by paulcurtis26** [savonet/liquidsoap#1847] [Q&A] [answered]
Hi there I have been trying for a while to get the Server functions for Liquidsoap working

```
def add_curtis_move_command(s) = 
  #add skip command
  def movetocurtisradio(skip) = 
  
   lines = get_process_lines("/home/pi/Music/NewMusicRadio/CopyFilesToCurtisRadio.sh")
   #"Copied"
   #"Moving from New Music to Curtis Radio"
   #skip = "false"
   if skip == "true" then
      log("skip is true")
      #source.skip(s)
    else
      log("skip is false")
    
     end
    end

#server.register(namespace="NewMusicRadio",usage="movetocurtisradio",description="Moves the current","curtiscopycmd",movetocurtisradio) - This did not work either
  server.register(namespace="NewMusicRadio",usage="movetocurtisradio",description="Moves the current","curtiscopycmd",fun (s) -> begin (s).movetocurtisradio("false") "Done" end)
  
end
```

when I run this I get a server command `NewMusicRadio.movetocurtisradio`
however when I run this command I get 

```
ERROR: unknown command, type "help" to get a list of commands.
END
```

I have spent hours changing the code/registration/functions and nothing gives me the server control.

do you have any good documentation on how to use the functions etc, as everything I find is 3 lines of code snippets without any context of how to use them

Thanks

Paul

**Comment by vitoyucepi** [comment, marked as answer]
Hi, @paulcurtis26.
I have the same problem at start. There's no sources over the internet with liquidsoap code samples.
The only way to learn the language is to read official wiki page about it.

Also I recommend to use the [book](https://www.liquidsoap.info/doc-dev/book.html) - it has some code snippets and other things.

And finally I created a local clone of official repo and look for examples there.
I've never used OCaml so the sources are not very useful for me, but I can definitely recommend [libs](https://github.com/savonet/liquidsoap/tree/main/libs) and [tests](https://github.com/savonet/liquidsoap/tree/main/tests).

As for your code. This is reformated snippet.
```
def addCurtisMoveCommand(s)
  def moveToCurtisRadio(skip)
    lines = get_process_lines("/home/pi/Music/NewMusicRadio/CopyFilesToCurtisRadio.sh")
    if skip == "true" then
      log("skip is true")
      source.skip(s)
    else
      log("skip is false")
    end
    "OK"
  end

  server.register(
    namespace="NewMusicRadio",
    usage="curtiscopycmd <boolean>",
    description="Moves the current",
    "curtiscopycmd",
    moveToCurtisRadio
  )
end
```
1. `server.register()` arguments
    - `usage` is how this will look in `help` command output.
      In this case I've added `<boolean>`, because I think you would expect to call your function like
      `NewMusicRadio.curtiscopycmd` + `true` or `false`.
      Maybe you should rename it like `usage="curtiscopycmd <true|false>"`.
    - Unlabeled argument `"curtiscopycmd"` is what command would server expect.
    - Unlabeled argument `moveToCurtisRadio` is callback or function that would execute after entering the server command.
1. `moveToCurtisRadio(skip)`
   It accepts only one argument - the rest of the command.
   `NewMusicRadio.curtiscopycmd true or abc` will call your function with argument `true or abc`.
   It must return a string, that will be displayed as a result to `telnet` user.
   In my code sample it returns `"OK"`, but you can add `lines` parsing and return `OK` or `FAIL`.
1. I'd prefer not to write long anonymous function because of readability issues.
